welcome-emails: Separate followup_day1 email from welcome emails.

laurynmm opened the pull request:
Separates the `followup_day1` email from the scheduled welcome emails, and instead sends the email directly when a new user account is created.

Fixes #25268.

**Notes**:
- There are 4 small prep commits with minor updates to tests and code comments.
- Did the changes to `followup_day1` email in two steps:
  - First, I removed it from `enqueue_welcome_emails`, creating a specific function for the `followup_day1` email and always scheduling it (without delay) when a new (human) user is created/processed.
  - Then, I switched it from being scheduled (without delay) to being sent via `send_email` in a separate commit. If we want to keep the email as a scheduled email, then it is easy to drop this commit, but it seemed cleaner for the implementation to make this change.
- I didn't change the email template name in these changes even though `followup_day1` is probably not the best name for this email.
  - Doing so will impact translations, and I wasn't sure how renaming the template might impact existing scheduled emails with the current template name (though they should be at the top of the send email queue).
  - I thought that renaming the template could be a follow-up task if people feel it's important to change. After these changes there are two remaining instances of `followup_day1` in the `zerver` directory: when we send the email as we reference the template name there, and a currently commented out translated string test in `test_i18n.py`.
- The final commit is an update to the help center article on disabling welcome emails in an organization.
- Note there is no change to the `followup_day1` email content / subject / sender / etc. in these changes.

**Questions**:
- Is this a change that we would want to inform organization administrators of?
  - I don't know how many Zulip Cloud organizations have sending welcome emails disabled, but I can see it being worth a heads up that we now send this email for all new user accounts created.
  - And noting it in the next major release for self-hosted organizations might make sense as well?

**Screenshots and screen captures:**

<details>
<summary>Help center: Disable welcome emails</summary>

[Current documentation](https://zulip.com/help/disable-welcome-emails)
![Screenshot from 2023-07-04 13-06-47](https://github.com/zulip/zulip/assets/63245456/95f8c334-908c-4d99-bd63-c4b466b15fa8)
</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented:
@alexmv - Would you be up for reviewing this when you get a chance?

alexmv commented:
> Then, I switched it from being scheduled (without delay) to being sent via send_email in a separate commit. If we want to keep the email as a scheduled email, then it is easy to drop this commit, but it seemed cleaner for the implementation to make this change.

I'm going to focus on this change, because it's an important one which I think is being glossed over.  From the commit message:

> **welcome-emails: Send followup_day1 email on account creation.**
> Instead of creating a scheduled email that has no delay for the
> followup_day1 email, which contains important account information,
> we now just send that email directly during account creation.
> 
> Since we send the followup_day1 email directly during account
> creation, we can remove it from the EMAIL_TYPES for scheduled
> emails.
> 
> This means that we make one less database query as we are no longer
> creating a ScheduledEmail for this email in send_future_email.
> 
> And this means that we have one more cache try when we get the user
> by ID for send_email.

Making one fewer database query, or doing one fewer row insert, isn't really a solid motivation for doing this.  This message really needs to motivate me as to if there are any serious tradeoffs here, and why this change is the right one.

One thing I don't know offhand is if this codepath is run inside of a transaction.  Hopefully it's not (for reasons clarified below) but whether it is or not should probably also make it into the commit message, because it makes a big difference.

Moving from being deferred (via the worker queue) to being done in-process has a few major effects:
 - Failures to send email will throw an exception at this point.  And email servers will, occasionally, just totally fall over and fail, so this _will happen_.  We do not retry syncronous email delivery attempts (unlike email via a queue), so this will be an immediate failure.
 
   Are we in a transaction?  If so, then failures to send the initial email will _fail the user creation process_.  If we're _not_ in a transaction, then failures to send this email show a 500 to the end-user but actually succeed in user creation, which may be a confusing experience (and we should handle via catching the exception, for instance).  Either of these are _possibly_ better than the current behaviour, where failure to send the email happens _asynchronously_ and _invisibly_ to this code (since it currently happens in the queue worker).  If sending the welcome email is a security-critical operation, then failures to send it should abort user creation.  My take is that it's probably-fine if sending the email fails, since it's not all that security critical, and end-users can't maliciously force such a failure (without taking down all of Amazon SES, which would be impressive).

 - Requests will take longer, because there's now a network call to the outside world involved, which we do not control.  If they take too long, the request may time out -- either in the client (brower getting bored, user pressing stop, etc) in which case the server's SMTP connection and other tasks will carry on, or in the server (by being terminated by uwgi for taking too long).  In the latter case, the process is unceremoniously terminated.
  
   If we're in a transaction, that rolls everything back (see above) -- _except things outside the database_.  So if the remote SMTP server accepted the email and was just hanging out for some reason, we've now sent email about a successful account creation _which we just rolled back_ and thus doesn't exist.  For this reason, email-sending (and interactions with any system outside of the database) is generally reserved for after the transaction has committed.

   If we're not in a transaction, then the user sees a 504 from nginx, and has the same questions as if they get the 500.

   In either case, the synchronous request taking longer could have knock-on effects.  This isn't a high-traffic endpoint, so in this case I'm not concerned -- but in general, it's worth thinking about how the system will deal with backpressure if the service it's calling (here, SMTP) starts to get slow as it gets overloaded.  If there's a queue worker in the way, then our queue will back up and may hammer on the service all the harder with automatic retries from the queue.  If the request is synchronous, then _our_ service will start to slow down as more django workers get tied up trying to talk to the busy service, and our end users will see the errors -- but since they're seeing errors, they might modulate their usage based on that feedback.  Again, not relevant to this low-traffic endpoint, just tossing it out as food for thought when thinking about (de)coupling systems.

 - If both welcome emails and you-have-an-account emails go out through the queue worker, they are _ordered_[^1].  Our current queue worker situation guarantees that delivery happen in the order they are inserted into the queue, so we can control the order in which they show in in people's inboxes.  If we switch one of them to being async, we not longer have any ordering guarantees.  _Most_ of the time the synchronous delivery will probably hit first, but that's not a sure thing.  So the question here is what order do we want, and how strongly do we care about that ordering?


Sorry for the long ramble!  I think on net, keeping this via the queue worker is probably the most sensible, but I'm open to discussion on it.

[^1]: This is a simplification.  They are ordered in the order that they are delivered to our outgoing SMTP server, which can (and sometimes does) reshuffle things in the order that it hands them off to intermediate or final mailservers.  So the order in which the remote end gets them inserted into their mailbox is never completely under our control.  However, we currently generate the `Date` header when we perform the handoff to the SMTP server, and most email clients order by that, so the order of SMTP handoff is _mostly_ predictive of the order they appear in clients mailboxes.

laurynmm commented:
@alexmv - Thanks for the thorough explanation of your concerns with no longer sending the "followup_day1" email via the scheduled email queue. Given that it makes sense to maintain the order of sending this email before any other "welcome email" that might be subsequently scheduled and that sending this email shouldn't impact the flow of a successful user/realm creation to logging the user into the (new) realm, I've dropped the commits that were directed at sending the email directly.

**Updates**:
- The "followup_day1" email is refactored to be separate for all other welcome emails that are scheduled and is sent independently of the organization's policy for `send_welcome_emails`. It is still scheduled via `send_future_email` without any delay.

alexmv commented on zerver/actions/create_user.py:
```diff
@@ -251,7 +251,11 @@ def process_new_human_user(
     clear_scheduled_invitation_emails(user_profile.delivery_email)
     if realm.send_welcome_emails:
-        enqueue_welcome_emails(user_profile, realm_creation)
+        enqueue_welcome_emails(user_profile)
+
+    # Schedule an initial email with the user's
+    # new account details and log-in information.
+    send_account_registered_email(user_profile, realm_creation)
```
Oh -- I think in my ramble previously, I thought that one of the welcome emails was also immediate -- but they're the day-2 (well, "day 2") and onboarding-zulip-guide emails, so my tangent about message ordering is totally irrelevant, since they're delayed by a while.  So never mind that part!

timabbott replied:
No I think you were correct -- the followup_days email is immediate, and this PR is recategorizing it as an account registration email. BTW @laurynmm we should finish the job of purging the misleading `followup_day1` name.

```
$ git grep followup_day1 | grep -v locale | wc
     10      31    1190
```

laurynmm replied:
Yeah, the ordering isn't super important with the current onboarding email schedule, but it's always possible that in the future a new onboarding email might be scheduled to be sent sooner than the current 2 days after registration.

For the rename of the email template, I'll do that as a follow up pull request. I think we also will want to take care of the "followup_day2" email rename in that same change. I'm thinking:
- followup_day1 -> account_registered
- followup_day2 -> onboarding_zulip_topics
- onboarding_zulip_guide, *no change*

@alexmv - Is there anything I need to be particularly aware of for that migration? For example, I'm thinking that we'd want to create new email templates (with the new names) and update the code that creates the scheduled emails to use those, then do a migration commit for existing scheduled messages that have the old template names to have the new ones, and then delete the old email templates in a final commit.

I'm not sure how that will affect the translated email strings, but it seems better to deal with extra updates in the translation process, and not have a situation where we're trying to send an email for a template that doesn't exist.

alexmv replied:
Yeah, you're absolutely spot on that the complexity is that the old names exist in the `ScheduledEmail` table, so we need a migration to update them.

In our production deploy, we'll need to stop the worker before starting the migration on staging, but that's not something that we can address in the general migration.

laurynmm replied:
Opened #26385 as a draft/WIP of this scheduled email template rename.

alexmv commented on zerver/lib/email_notifications.py:
```diff
@@ -756,52 +756,51 @@ def get_org_type_zulip_guide(realm: Realm) -> Tuple[Any, str]:
-        # line break to avoid triggering lint rule
-        from_name = settings.WELCOME_EMAIL_SENDER["name"]
-        from_address = settings.WELCOME_EMAIL_SENDER["email"]
+        from_name = str(settings.WELCOME_EMAIL_SENDER["name"])
+        from_address = str(settings.WELCOME_EMAIL_SENDER["email"])
```
I don't think we ned the `str()` calls here.

alexmv reviewed (commented):
This is great, @laurynmm!  One tiny nit, but otherwise LGTM.

timabbott commented on help/disable-welcome-emails.md:
```diff
@@ -2,10 +2,12 @@
+your organization, you can disable them.
+
+Note that new users will receive an initial email to inform them about
+their new Zulip account and how to log in.
```
Maybe clarify more explicitly that this is independent of that setting, e.g., "Note that regardless of this setting, users will receive an initial account email notifying them about their new Zulip account and what URL to use to log in."

laurynmm commented:
**Updates**:
- Removed unnecessary `str()` calls in new `welcome_sender_information` helper function.
- Added clarification to help center documentation update to **Disable welcome emails** article, see updated screenshot below.

**Follow-ups**:
- Rename the "followup_day1" and "followup_day2" email templates to have names that more clearly reflect the email content and purpose. See [this comment](https://github.com/zulip/zulip/pull/26182#discussion_r1259683762).

**Updated screenshot**:

<details>
<summary>Disable welcome emails</summary>

[Current documentation](https://zulip.com/help/disable-welcome-emails)
![Screenshot from 2023-07-11 14-33-11](https://github.com/zulip/zulip/assets/63245456/6efc80fe-9863-43a9-97a7-26f231dcf8e8)
</details>

timabbott commented:
Merged, thanks @laurynmm!
